server-protocol.md

```diff
@@ -92,6 +85,26 @@ following keys (and ignores all others):
   future version of the protocol requires a rate-limiting CAPTCHA ticket or
   other authorization record, the server can send `error` (explaining the
   requirement) if it does not see this ticket arrive before the `bind`.
+* `permission-required`: a set of available authentication methods,
+  proof of work challenges etc. The client needs to "solve" one of
+  them in order to get access to the service.
+
+The client should examine the `permissions-required` methods (if
+present) and select one to use (see also "Permission to Use the
+Server" below). If the client doesn't support any permissions methods
+or has already chosen not to use any then it _may_ send `bind`
+immediately upon connecting. Older clients that do not know about
+`permsision-required` will always send the `bind` immediately (without
+waiting for the `welcome` message). In that case, the server MAY grant usage permission 
+regardless. Otherwise, it will send a "permission denied" error message and close the connection.
+
+Clients that do support Permission and select a method other than `none` must send a `submit-permissions` message before a `bind` message.
+No matter when it is sent, the `bind` messages specifies the AppID and
```

Comment:
Please specify an arbitrary but fixed position where the `bind` message may occur if present. Having the control flow as fixed as possible makes the protocol simple and implementations easier.

Personally, I suggest `bind` after `submit-permissions`, because I think the permissions should be scoped to the whole WebSockets connection. This also applies to "late" `bind` messages: sending them after "normal", unbound messages have been exchanged should be illegal as well (this is not necessarily worth an explicit mention in the protocol text depending on the exact wording).

So the intended order is `welcome` > `submit-permissions` > `bind` (the latter two being individually optional), but `bind` > `welcome` must be supported by the server for backwards compatibility.

Reply:
Yes, `submit-permissions` before `bind`. The "no matter when it is sent" phrasing was to encompass both ways: an "older" client that sends `bind` immediately, or a newer one that sends it after waiting for the `welcome` (and then maybe also sends a `submit-permissions` first).